### ISSUE TITLE
Folded Ref Structs with Fixed Buffer causes System.AccessViolationException.

### ISSUE THREAD

**Issue by VitaliiBachulis** [dotnet/roslyn#67799]
**Version Used**:  C# 11

**Steps to Reproduce**:

1.  Use source code in "Investigation" project.
2.  https://github.com/VitaliiBachulis/Investigation
3. In the ref struct constructor of outer struct has to initialize inner struct instead of creating a temporary one and copy it in a field.
Compiler has to generate:

```
  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call     instance void IntruderAlert.DebounceMeasurement::.ctor()
    IL_0006:  ret
  } // end of method Room::.ctor

instead of:

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       12 (0xc)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  newobj     instance void IntruderAlert.DebounceMeasurement::.ctor()
    IL_0006:  stfld      valuetype IntruderAlert.DebounceMeasurement IntruderAlert.Room::debounce
    IL_000b:  ret
  } // end of method Room::.ctor
```

**Comment by jcouv** [comment]
Might you be able to extract a smaller repro for your situation (preferably on [sharplab](https://sharplab.io/))? It sounds like you need two ref structs but I couldn't figure out what's needed beyond that. Thanks

**Comment by jcouv** [comment]
@VitaliiBachulis I can see the IL generated for the `Room()` constructor, but it looks like what I would have expected (constructs a new `DebounceMeasurement` then assigns it to the field). In terms of language rules, why should this not be the case?
Tagging @cston to advise (does the IL look wrong?).

**Comment by VitaliiBachulis** [comment]
If you use <TieredCompilation>false</TieredCompilation> - the project does not crashe and looks like all right. But the memory layout of  the resulting struct is doubled. And generaded ASM code is strange. After apply IL fix by ILDasm/ILAsm - the program works as expected - and no struct doubles in memory (approved by WinDbg).

**Comment by VitaliiBachulis** [comment]
The semantic meaning of folded ref structs - not instantiation of it's fields (ref struct) on the evaluation stack - but it's initialization.
Such as allocation of the outer struct has already just happened on it's declaration.

**Comment by VitaliiBachulis** [comment]
If you analyze IL code generated for instantiation of the Room struct. We can see that it generated - what is expected.
.method private hidebysig static 
	void '<Main>$' (
		string[] args
	) cil managed 
{
	// {
	IL_0000: ldloca.s 0
	// new Room().TakeMeasurements(14);
	IL_0002: call instance void IntruderAlert.Room::.ctor()
	IL_0007: ldloca.s 0
	IL_0009: ldc.i4.s 14
	IL_000b: call instance void IntruderAlert.Room::TakeMeasurements(int32)

**Comment by VitaliiBachulis** [comment]
but the constructor of the Room struct does not do the same initialization logic in it's implementation.
.method public hidebysig specialname rtspecialname 
	instance void .ctor () cil managed 
{
	// Method begins at RVA 0x2126
	// Header size: 1
	// Code size: 12 (0xc)
	.maxstack 8

	// debounce = new DebounceMeasurement();
	IL_0000: ldarg.0
	IL_0001: newobj instance void IntruderAlert.DebounceMeasurement::.ctor()
	IL_0006: stfld valuetype IntruderAlert.DebounceMeasurement IntruderAlert.Room::debounce

**Comment by VitaliiBachulis** [comment]
sharplab does not allow to understand the problem. I have spend a lot of time using ILDasm and WinDbg to see what happens.

**Comment by VitaliiBachulis** [comment]
Value types are not usually created using newobj. They are usually allocated either as arguments or local variables, using newarr (for zero-based, one-dimensional arrays), or as fields of objects. Once allocated, they are initialized using [Initobj](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.initobj?view=net-7.0). However, the newobj instruction can be used to create a new instance of a value type on the stack, that can then be passed as an argument, stored in a local, and so on.

[newobj](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.newobj?view=net-7.0)

I think that for the Value types the newobj is acceptable in case of creating and than boxing a value type. ?

**Comment by VitaliiBachulis** [comment]
At first time I was fooled that problem is in a JIT compiler (because program crashed only in QuickJit and worked fine in Optimized compilation). But after apply an IL fix - I checked all variants and in all cases JIT generated ASM correctly and program works fine.

**Comment by VitaliiBachulis** [comment]
For really see the picture the best way is to look at virtual addresses values on an evaluation stack. I did step by step ASM's commands running and watched the result of commands on the stack.

**Comment by VitaliiBachulis** [comment]
.

**Comment by VitaliiBachulis** [comment]
> @VitaliiBachulis I can see the IL generated for the `Room()` constructor, but it looks like what I would have expected (constructs a new `DebounceMeasurement` then assigns it to the field). In terms of language rules, why should this not be the case? Tagging @cston to advise (does the IL look wrong?).

https://github.com/dotnet/roslyn/issues/67799#issuecomment-1509722414

**Comment by VitaliiBachulis** [comment]
I think this case could be included in an test plan for: #https://github.com/dotnet/roslyn/issues/67826

**Comment by cston** [comment]
I believe the current behavior is by design. The instance field initializer corresponds to an assignment to the field (see [language specification](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#15561-general)), and since the field is a `struct`, the initializer value is copied. As a result, the `Span<>` in the copied value refers to the `fixed` buffer in the original instance.

**Comment by jaredpar** [comment]
> If the current behavior is by design - how do you explain the next: https://github.com/dotnet/roslyn/issues/67799#issuecomment-1509705054 and https://github.com/dotnet/roslyn/issues/67799#issuecomment-1509706801

The compiler will in some cases initialize `struct` values in place. This is an optimization though, it is **not** a semantic guarantee. The compiler can, will and has changed when this optimization does and does occur through the lifetime of the compiler. 

> I think this case could be included in a test plan for: #67826

I agree this is a case we need to consider when thinking about inline arrays. The inline array case though is very different because it's entirely safe code. It is up to the compiler to ensure that the code doesn't end up referring to temporary artifacts. I'm pretty sure the existing lifetime rules account for this but it's a good case to make sure we test. 

The case you present here is very different. It's using `unsafe` code to create a `Span<T>`. That pattern is only safe if the `struct` is never copied. That is a _very_ hard guarantee to make with `struct` values in .NET because copying is fairly inherent to their nature. 

This particular behavior is "By Design" in that we do not guarantee, nor do we wish to guarantee, that copies do not happen in the `x = new()` case.
